Add details of ‘confirm email address’ and ‘custom retention period’ (#215)
alphagov/notifications-python-client: DOCUMENTATION.md

To send a file by email, add a placeholder to the template then upload a file. The placeholder will contain a secure link to download the file.

The links are unique and unguessable. GOV.UK Notify cannot access or decrypt your file.

> Your file will be available to download for a default period of 78 weeks (18 months). From 29 March 2023 we will reduce this to 26 weeks (6 months) for all new files. Files sent before 29 March are not affected.

Reviewer: Very small one that is optional as maybe our documentation doesn't need to be 100% precise necessarily.

Is it beneficial to say the default period is 78 weeks. Firstly, I'm guessing it may not always be exactly 78 weeks depending on how many weeks in a month? Secondly, at this point is it useful to make them start thinking in terms of weeks?

Same with the 26 weeks (6 months).

@samuelhwilliams Here is a slightly more interesting one though, what will our default retention period actually be when we change it. Will it be 26 weeks or will it be 6 months? Those could be very slightly different by a few days.

Reviewer 2: It'll be 26 weeks. Everything will be in weeks - which I think is why we're starting to get them to think that way by how we speak in the docs.

Reviewer 2: At the moment, documents in the S3 bucket expire after 546 days, which is 78 weeks. So that is correct.

Author: Thanks both, I’m resolving this.

To help protect your files you can also:

> * ask recipients to confirm their email address before downloading

Reviewer: Optional: will this sentence need changing in 6 months time? Maybe in 6 months time I'd phrase it more like we ask recipients to confirm their email address before downloading. You can choose to turn this off. 

I think for speed of getting this merged, I think we don't make a change on this for the moment at least.

Author: Yes, the whole section will need a little makeover in 6 months.

All the ‘turn it on’ stuff will go away, the retention period will be fixed at 26 weeks, etc.

* choose the length of time that a file is available to download

> To turn these features on or off, you will need to update your API client to the latest version.

Reviewer: I would suggest we remove this sentence for two reasons

1. We don't have this precedent anywhere else of telling people they need to update their API client to take advantage of a new feature
2. At some point it means we need to remove this line as it won't be true (as users by that point may already be using a version of the client that contains this functionality)

Reviewer 2: Alternatively we say that they need to update to at least a specific client version x.y.z.

    }
```

#### Ask recipients to confirm their email address before they can download the file

> This feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.

Author: ```suggestion
This new security feature is optional. You should use it if you send files that are sensitive - for example, because they contain personal information about your users.

When a recipient clicks the link in the email you’ve sent them, they have to enter their email address. Only someone who knows the recipient’s email address can download the file.

From 29 March 2023, we will turn this feature on by default for every file you send. Files sent before 29 March are not affected.
```

Reviewer 2: "won't be affected" vs "are not affected"?

Otherwise looks good.

> ##### Opt in

Author: ```suggestion
##### Turn on email address check
```

> To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.
> 
> You will not need to do this after 29 March 2023.

Author: ```suggestion
You can turn this feature on on a per-file basis. You will not need to do this after 29 March 2023.

To make the recipient confirm their email address before downloading the file, set the `verify_email_before_download` flag to `true`.
```

Reviewer 2: Do we drop off the '2023' second time around just because it's too repetitive?

```python
from notifications_python_client import prepare_upload

with open('file.pdf', 'rb') as f:
    ...
    personalisation={
      'first_name': 'Amala',
      'application_date': '2018-01-01',

>       'link_to_file': prepare_upload(f),

Author: ```suggestion
      'link_to_file': prepare_upload(f),
      'verify_email_before_download': true,
```

    }
```

> ##### Opt out (not recommended)

Author: ```suggestion
##### Turn off email check (not recommended)
```

> You should not opt out if you send files that contain personally identifiable information or sensitive information.

Author: ```suggestion
You can turn this feature off on a per-file basis. You should not do this if you send files that contain either:

* personally identifiable information
* commercially sensitive information
```

Reviewer: Same comment about updating the API client as above

Reviewer: I'm interested in why we decided to include `information classified as ‘OFFICIAL’ or ‘OFFICIAL-SENSITIVE’ under the Government Security Classifications policy` please? Was this a suggestion from IA or privacy?

And should we link to the Government Security Classifications policy here?

> If you do not want to use this security feature, set the `verify_email_before_download` flag to `false` by 29 March 2023.

Author: ```suggestion
```

```python
from notifications_python_client import prepare_upload

with open('file.pdf', 'rb') as f:
    ...
    personalisation={
      'first_name': 'Amala',
      'application_date': '2018-01-01',

>       'link_to_file': prepare_upload(f),

Author: ```suggestion
      'link_to_file': prepare_upload(f),
      'verify_email_before_download': false,
```

Reviewer 2: Same here:

```suggestion
      'link_to_file': prepare_upload(f),
      'verify_email_before_download': False,
```

    }
```

#### Choose the length of time that a file is available to download

Set the number of weeks you want the file to be available using the `retention_period` key.

You can choose any value between 1 week and 78 weeks.

If you do not do this, the file will be available for the default period of 78 weeks (18 months).

```python
from notifications_python_client import prepare_upload

with open('file.pdf', 'rb') as f:
    ...
    personalisation={
      'first_name': 'Amala',
      'application_date': '2018-01-01',

>       'link_to_file': prepare_upload(f),

Author: ```suggestion
      'link_to_file': prepare_upload(f),
      'retention_period': '4 weeks',
```

Reviewer 2: And here:

```suggestion
      'link_to_file': prepare_upload(f, retention_period='4 weeks'),
```

    }
```

#### Ask recipients to confirm their email address before they can download the file

This feature is currently opt-in only. From 29 March 2023 it will apply to all new files by default, unless you opt out.

##### Opt in

To ask recipients to confirm their email address, set the `verify_email_before_download` flag to `true`.

You will not need to do this after 29 March 2023.

```python
from notifications_python_client import prepare_upload

with open('file.pdf', 'rb') as f:
    ...
    personalisation={
      'first_name': 'Amala',
      'application_date': '2018-01-01',

>       'link_to_file': prepare_upload(f),
>       'verify_email_before_download': True,

Author: ```suggestion
      'link_to_file': prepare_upload(f, verify_email_before_download=True),
```

    }
```

##### Turn off email address check (not recommended)

If you do not want to use this feature after 29 March 2023, you can turn it off on a file-by-file basis.

To do this you will need version X.X.X of the Python client library, or a more recent version.

You should not turn this feature off if you send files that contain:

* personally identifiable information
* commercially sensitive information
* information classified as ‘OFFICIAL’ or ‘OFFICIAL-SENSITIVE’ under the [Government Security Classifications](https://www.gov.uk/government/publications/government-security-classifications) policy

To let the recipient download the file without confirming their email address, set the `verify_email_before_download` flag to `False`.

```python
from notifications_python_client import prepare_upload

with open('file.pdf', 'rb') as f:
    ...
    personalisation={
      'first_name': 'Amala',
      'application_date': '2018-01-01',
      'link_to_file': prepare_upload(f, verify_email_before_download=False),
    }
```

#### Choose the length of time that a file is available to download

Set the number of weeks you want the file to be available using the `retention_period` key.

> You can choose any value between 1 week and 78 weeks.
> 
> If you do not do this, the file will be available for the default period of 78 weeks (18 months).
> 

Author: ```suggestion
You can choose any value between 1 week and 78 weeks.

To use this feature will need version X.X.X of the Python client library, or a more recent version.

If you do not choose a value, the file will be available for the default period of 78 weeks (18 months).

```
